Add mqtt auth server in go (#786)
Azure/azure-webpubsub: samples/golang/mqttClientCertAuthServer/README.md

# MQTT Client Authentication and Authorization Server Based On Client Certificate

This project is an server to authenticate and authorize MQTT client based on client certificate.

-

## Running the server
To run the server, follow these simple steps:

```
go run main.go
```

To run the server in a docker container
```
docker build --network=host -t openapi .
```

Once image is built use
```
docker run --rm -it openapi
```

## Contributing guidance

This server is based on a generated version by the [openapi-generator]
(https://openapi-generator.tech) project from [Swagger API spec](..\..\..\\protocols\server\cloud-events\tsp-output\@typespec\openapi3\openapi.yaml).

After installing the Open API generator, you can generate the code with the following command:
```bash
openapi-generator-cli generate -i ..\..\..\\protocols\server\cloud-events\tsp-output\@typespec\openapi3\openapi.yaml -g go-server -o .
```

The following files are customized:
1. In [api_default_service.go](./go/api_default_service.go), func `OnMqttConnect` is updated to show how to auth clients based on client certificates.

> 2. In [api_default.go], the generated code `d.DisallowUnknownFields()` to disallow unknwon properties is already made into comment. **As it's expected to have new properties added in the Web PubSub CloudEvents protocol, you MUST allow unknown properties when deserializing request payload from Web PubSub.**

Reviewer: ```suggestion
2. In [api_default.go], `d.DisallowUnknownFields()` is added into comment. **As it's expected to have new properties added in the Web PubSub CloudEvents protocol, you MUST allow unknown properties when deserializing request payload from Web PubSub.**
```

> 3. In [go.mod](./go.mod), required dependencies are added.

Reviewer: also highlight the added dependencies? is it related to the code added in `OnMqttConnect` or they are always needed?

Author: They are always needed. I think it's bug that it's not included in the generation code.

# MQTT Client Authentication and Authorization Server Based On Client Certificate

This project is an server to authenticate and authorize MQTT client based on client certificate.

-

## Running the server
To run the server, follow these simple steps:

```
go run main.go
```

To run the server in a docker container
```
docker build --network=host -t openapi .
```

Once image is built use
```
docker run --rm -it openapi
```

## Contributing guidance

This server is based on a generated version by the [openapi-generator]
(https://openapi-generator.tech) project from [Swagger API spec](..\..\..\\protocols\server\cloud-events\tsp-output\@typespec\openapi3\openapi.yaml).

After installing the Open API generator, you can generate the code with the following command:
```bash

> openapi-generator-cli generate -i ..\..\..\\protocols\server\cloud-events\tsp-output\@typespec\openapi3\openapi.yaml -g go-server -o .

Reviewer: could it generate the files into a `_generated` folder instead?

```

> The following files are cutomized:

Reviewer: ```suggestion
The following files are customized:
```

> 1. [api_default_service.go](./go/api_default_service.go) contains the business logic to auth clients based on client certificates.

Reviewer: list the changes by the file name?
```suggestion
1. In [api_default_service.go](./go/api_default_service.go), func `OnMqttConnect` is updated to show how to auth clients based on client certificates.
2. In [api_default.go]: 
    1. change1
    2. change2
```

2. The [generated code](./go/api_default.go) to disallow unknwon properties is already made into comment. **As it's expected to have new properties added in the Web PubSub CloudEvents protocol, you MUST allow unknown properties when deserializing request payload from Web PubSub.**

> 1. The [route()](./go/api_default.go) function is updated to modify exposed endpoints.

Reviewer: is route() modification still needed when the typespec is updated?

Author: Updated.

3. [go.mod](./go.mod) is updated to add required dependencies.

> 4. All the above files are added into[.openapi-generator-ignore](./.openapi-generator-ignore) so that regeneration won't override the cutomization part.

Reviewer: this ignore will skip the entire file right? Any case the regeneration needs to update the model in the ignored files?

Author: Those files don't contain models.
